packages/plugins/other/visitor-plugin-common/src/selection-set-to-object.ts

```diff
@@ -102,7 +102,7 @@ export class SelectionSetToObject<Config extends ParsedDocumentsConfig = ParsedD
           this._appendToTypeMap(types, typeOnSchema.name, fields);
           this._appendToTypeMap(types, typeOnSchema.name, spreadsUsage[typeOnSchema.name]);
           this._collectInlineFragments(typeOnSchema, inlines, types);
-        } else if (isInterfaceType(typeOnSchema) && parentType.isTypeOf(typeOnSchema, null, null)) {
+        } else if (isInterfaceType(typeOnSchema) && parentType.getInterfaces().includes(typeOnSchema)) {
```

Comment:
@benmosher Could you please add a test case for the functionality? In general for an operation that spreads a fragment on a type where it does not belong.

Also since interfaces can implement interfaces, it seems as we would also have to use recursion as the getInterfaces function on parentType could also return an interface that implements another interface.

```graphql
interface A {
  a: String!
}

interface B implements A {
  a: String!
  b: String
}

type C implements B {
  a: String!
  b: String
  c: String!
}

type Query {
  foo: C
}
```

```graphql
query {
  foo {
    ... on A {
      a
    }
  }
}
```

Reply:
I will if I get a chance, but I'm not sure when that will be. also, is this syntax valid? I recently added an interface to another interface and got a warning in my editor that was essentially `Type 'C' must also implement A` and had to update my analogous concrete types to read as `type C implements B & A {`.

I believe the specification bears this out: https://github.com/graphql/graphql-spec/pull/373/files#diff-30a69c5a5eded8e1aea52e53dad1181e6ec8f549ca2c50570b035153e2de1c43R1090

Reply:
@benmosher That's fair! I wasn't aware of that detail.

I will see whether I can add a test for the spread tomorrow :)

Reply:
I added this test. The output seems fine: `{ foo?: Maybe<{ __typename?: 'C' }> }`. GraphQL allow you to spread on incorrect types, so that's fine :) 
Also, did a rebase for this PR. Looks great and we can go ahead and merge it.